OCaml_Intro_to_CS/part_6.ml

```diff
@@ -0,0 +1,220 @@
+          )
+        else 
+          (
+            print_endline "Applying function from list with acc";
+            aux (h acc) t
+          )
+  in
+  aux 0 lst
+;;
+
+let operations = [inc ; double ; dec];;
+
+let res = compose_list operations 7;;
+
+let () = 
+  Printf.printf "Result of compose_list operations: %d\n" res;;
+
+(* for the extra 1 mark, using fold_left and fold_right *)
+let compose_list_fl func_lst x =
+  let f acc func = func acc in
+  let init = x in
+  List.fold_left f init func_lst
+;;
+
+let fl_res = compose_list_fl operations 7;;
+
+let () = 
+  Printf.printf "Result of compose_list_fl operations: %d\n" fl_res;;
+
+let compose_list_fr func_lst x =
+  let f func acc = func acc in
+  let init = x in
+  List.fold_right f func_lst init
+;;
+
+let fr_res = compose_list_fr operations 7;;
+
+let () = 
+  Printf.printf "Result of compose_list_fr operations: %d\n" fr_res;;
+
+(*
+Your task: implement mapi : (int -> 'a -> 'b) -> 'a list -> 'b list.
+  *)
+
+let mapi f lst =
+  let mapper i acc x =
+    let mapped_value = f i x in
+    mapped_value :: acc
+  in
+  let indexed_list = List.mapi (fun i x -> (i, x)) lst in
+  let init = [] in 
+  let mapped_values = List.fold_left (fun acc (i, x) -> mapper i acc x) init indexed_list in
+  List.rev mapped_values
+;;
+
+let mapi_res = mapi (fun i a -> if i mod 2 = 0 then a + 1 else a - 1) [1;2;3;4;5];;
+
+let print_list lst = 
+  let str = String.concat ", " (List.map string_of_int lst) in
+  Printf.printf "[%s]\n" str
+;;
+
+print_endline "Result for mapi: ";;
+print_list mapi_res;;
+(*print_int mapi_res;;*)
+
+(*
+(2 marks, plus 1 mark of extra credit) Suppose we wish to generate a decent-sized test for the
+complex parenthesis matching problem. In particular we want a function generate_tests :
+int -> char list list that generates all of the tests of length 2n (where n is the input
+argument) that have the following structure. The first n characters should be open parentheses,
+and the last n characters should be the appropriate matching closed parentheses. For example,
+here is the output for my function when n=3:
+*)
+
+let combinations n = 
+  let rec aux acc k =
+    if k = 0 then acc
+    else
+      aux (List.concat [
+        List.map (fun l -> '(' :: l) acc;
+        List.map (fun l -> '{' :: l) acc;
+        List.map (fun l -> '[' :: l) acc
+      ]) (k - 1)
+    in
+    aux [[]] n
+  ;;
+
+let matching_parenthesis open_parenthesis =
+  List.map (function
+  | '(' -> ')'
+  | '{' -> '}'
+  | '[' -> ']'
+  | _ -> failwith "Invalid character"
+  ) open_parenthesis
+;;
+
+
+let generate_tests n =
+  (* first to have a list of all brackets, then run through the various permutations of the brackets in a BFS way *)
+  (* in python it will be something like def backtrack(temp, list, result) *)
+  let open_combinations = combinations n in
+  List.map (fun open_parenthesis ->
+    open_parenthesis @ (matching_parenthesis open_parenthesis)
+  ) open_combinations
+  ;;
```

Comment:
Yup, this is how I would do it too.

Reply:
Im actually dead. I had to scratch my head for 10 years

Reply:
But good though, It's not easy to move from thinking about programming procedurally to functionally and you're doing well